loopy/kernel/__init__.py

```diff
@@ -151,7 +151,24 @@ def _get_inames_from_domains(domains):
             (frozenset(dom.get_var_names(dim_type.set)) for dom in domains))
 
 
-class LoopKernel(ImmutableRecordWithoutPickling):
+class ImmutableRecordWithoutPickingWithTargetedCopies(ImmutableRecordWithoutPickling):  # noqa: E501
+    def __getattr__(self, key):
+        if key.startswith("with_"):
+            skey = key[5:]
+            if skey in self.fields:
+                return lambda x: ImmutableRecordWithoutPickling.copy(self,
+                                                                     **{skey: x})
+
+        return super().__getattr__(key)
+
+    def copy(self, **kwargs):
+        from functools import reduce
+        return reduce(lambda obj, kwarg: getattr(obj,
+                                                 f"with_{kwarg[0]}")(kwarg[1]),
+                      kwargs.items(), ImmutableRecordWithoutPickling.copy(self))
```

Comment:
I'm not a huge fan of this. This goes through the (to be fair: new, leaner) constructor once per changed attribute. I think we'll come to regret that cost. Why not do the invalidation processing in `get_copy_kwargs`?

Reply:
`Record.get_copy_kwargs`[[1]](https://github.com/inducer/pytools/blob/c8352215742fc454946818a183e66afe8625e40d/pytools/__init__.py#L367) get called every time `Record.copy` gets called. So even that would lead to attribute checking once every attribute being copied.

I am slightly against doing the invalidation processing separately somewhere to avoid code duplication.

Reply:
Moved the implementation to `get_copy_kwargs` in 0463041bdd0775ed10ef1b2ef07c93e9d5d836fc.